OpenRA.Mods.Common/Widgets/SpritePowerMeterWidget.cs

```diff
@@ -0,0 +1,119 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using OpenRA.Graphics;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public class SpritePowerMeterWidget : Widget
+	{
+		[Desc("Height of each meter bar")]
+		[FieldLoader.Require]
+		public readonly int MeterHeight = 4;
+
+		[Desc("How many units of power each bar represents.")]
+		[FieldLoader.Require]
+		public readonly int PowerUnitsPerBar = 25;
+
+		[Desc("Blank Image for the meter bar")]
+		[FieldLoader.Require]
+		public readonly string NoPowerImage = "";
+
+		[Desc("When you have access power to use.")]
+		[FieldLoader.Require]
+		public readonly string AvailablePowerImage = "";
+
+		[Desc("Used power image")]
+		[FieldLoader.Require]
+		public readonly string UsedPowerImage = "";
+
+		[Desc("Too much power used meter image.")]
+		[FieldLoader.Require]
+		public readonly string OverUsedPowerImage = "";
+
+		[Desc("The power animation sprite image.")]
+		[FieldLoader.Require]
+		public readonly string FlashPowerImage = "";
+
+		[Desc("The collection of images to get the meter images from.")]
+		[FieldLoader.Require]
+		public readonly string ImageCollection = "";
+
+		[Desc("How long to blink on low power.")]
+		[FieldLoader.Require]
+		public readonly int WarningFlashDuration = 10;
+
+		[Desc("How fast to blink on low power.")]
+		[FieldLoader.Require]
+		public readonly int WarningFlashBlinkRate = 80;
+
+		public Widget SidebarProduction;
+
+		public decimal NumberOfBars;
+
+		public int WarningFlash;
+		public int TotalPowerDisplay;
+		public int LastTotalPowerDisplay;
+
+		public decimal TotalPowerStep;
+		public decimal PowerUsedStep;
+		public decimal PowerAvailableStep;
+
+		public bool LowPower;
+
+		[ObjectCreator.UseCtor]
+		public SpritePowerMeterWidget() { }
+
+		public override void Draw()
+		{
+			NumberOfBars = decimal.Floor(Bounds.Height / MeterHeight);
+
+			// Create a list of new bars
+			for (int i = 0; i < NumberOfBars; i++)
+			{
+				var image = NoPowerImage;
+
+				var targetIcon = AvailablePowerImage;
+
+				if (i < PowerUsedStep)
+					targetIcon = UsedPowerImage;
+
+				if (i > PowerAvailableStep)
+					targetIcon = OverUsedPowerImage;
+
+				if (i == TotalPowerStep && LowPower)
+					targetIcon = OverUsedPowerImage;
+
+				// Flash the top bar if something is wrong
+				if (i == TotalPowerStep)
+				{
+					if (WarningFlash % WarningFlashBlinkRate != 0)
+						targetIcon = FlashPowerImage;
+					if (WarningFlash > 0)
+						WarningFlash--;
+				}
+
+				// We exit if updating a bar meter. This gives a nice animation effect
+				if (image != targetIcon)
+					image = targetIcon;
+
+				var sprite = ChromeProvider.GetImage(ImageCollection, image);
```

Comment:
This has a non-trivial performance overhead, so really shouldn't be done 60 * (number of bars) times per second. Cache `Sprite availablePowerImage` etc from the widget Init and then use them directly in this loop.